noisy-simulator/src/density_matrix_simulator.rs

```diff
@@ -0,0 +1,346 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+//! This module contains two structs: the `DensityMatrixSimulator` and its
+//! internal `DensityMatrix` state.
+
+#[cfg(test)]
+mod tests;
+
+use crate::{
+    handle_error, instrument::Instrument, kernel::apply_kernel, operation::Operation,
+    ComplexVector, Error, NoisySimulator, SquareMatrix, TOLERANCE,
+};
+use num_complex::Complex;
+use rand::{rngs::StdRng, Rng, SeedableRng};
+
+/// A vectorized density matrix.
+#[derive(Debug, Clone)]
+pub struct DensityMatrix {
+    /// Dimension of the matrix. E.g.: If the matrix is 5 x 5, then dimension is 5.
+    dimension: usize,
+    /// Number of qubits in the system.
+    number_of_qubits: usize,
+    /// Theoretical change in trace due to operations that have been applied so far.
+    trace_change: f64,
+    /// Vector storing the entries of the density matrix.
+    data: ComplexVector,
+}
+
+impl DensityMatrix {
+    fn new(number_of_qubits: usize) -> Self {
+        let dimension = 1 << number_of_qubits;
+        let mut data = ComplexVector::zeros(dimension * dimension);
+        data[0].re = 1.0;
+        Self {
+            dimension,
+            number_of_qubits,
+            trace_change: 1.0,
+            data,
+        }
+    }
+
+    /// Builds a `DensityMatrix` from its raw fields. Returns `None` if
+    ///  the provided args don't represent a valid `DensityMatrix`.
+    ///
+    /// This method is to be used from the PyO3 wrapper.
+    pub fn try_from(
```

Comment:
The try_from pattern uses `Result<T, E>` whereas this is using `Option<T>`.

Reply:
Fixed